### ISSUE TITLE
Support export with grayscale input mode

### ISSUE THREAD

**Issue by talcs** [ultralytics/ultralytics#22960]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

It will be great if a `grayscale=True` option (or similar) is added to the model export API. It will need to replace the first conv layer to one that will have 1 channel instead of 3, and all kernels will be the channel-wise average of the learned ones. 

### Use case

The use case is on edge devices. Usually images are represented as `HxWx3`, while the network expects `3xHxW`. Swapping axes could be a costly operation on edge devices (rearranging the memory). Meanwhile, opting for grayscale input eliminates this problem, since swapping from `HxWx1` to `1xHxW` does not require any memory rearrangement. Being able to use a BGR-trained network that was automatically converted to accepting Grayscale inputs could be a great option.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @talcs, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Thanks for the thoughtful feature request around supporting grayscale input during export (e.g., a `grayscale=True` option that adapts the first conv layer) 🧠🔧 This is an automated response, and an Ultralytics engineer will also assist soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

To help us evaluate this export enhancement request faster, could you also share:-Which export format(s) you’re targeting (e.g., ONNX, TensorRT, CoreML, etc.)-Which YOLO model/task you’re using (detect/segment/classify/pose/obb)-What your current edge-device pipeline looks like (input tensor layout expectations, pre/post-processing constraints)-If you have any benchmarks showing the axis-swap overhead vs. grayscale input on your target hardware

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Why not train a grayscale model instead? Wouldn't this averaging approach decrease the inference accuracy?

**Comment by talcs** [comment]
Because when you feed a grayscale image to a BGR network, you simply duplicate the gray channel 3 times. You can simulate the inner product result exactly, by summing the channelwise weights of the first conv layer, while only feeding a single channel input to the net.

The coefficients should actually all be 1 rather than the sample code's 0.3333.

**Comment by Y-T-G** [comment]
But if your model is intended for grayscale, you can just train a grayscale model and it will also export with single channel input:

https://docs.ultralytics.com/datasets/detect/coco8-grayscale/

**Comment by talcs** [comment]
In most cases people reuse pretrained models rather than training from scratch. The trick I'm proposing can reduce the preprocessing time significantly, while it doesn't require training anything. It just ports a trained BGR model to grayscale instantly. Training can be a mess and can end up in an inferior model.

**Comment by Y-T-G** [comment]
But if you train the model on grayscale, wouldn't it be better at inferring on grayscale than training a model on RGB and then running inference on grayscale?

Is the original model you're converting trained on RGB or grayscale?

**Comment by talcs** [comment]
The original model is the Ultralytics pretrained YOLO model, trained on RGB of course.

I want to train nothing, just take a pretrained model and convert its first layer to accept grayscale. The grayscale first layer will prevent the grayscale->RGB conversion on edge devices (channel duplication), followed by the even more costly operation of swapping axes from HxWx3 to 3xHxW.

**Comment by Y-T-G** [comment]
I see. Makes sense.

You can manually update the Conv in `model.model.model[0]` to single channel and run export and it should work.

```
In [2]: model.model.model[0]
Out[2]:
Conv(
  (conv): Conv2d(3, 16, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
  (bn): BatchNorm2d(16, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
  (act): SiLU(inplace=True)
)

In [3]: model.model.model[0].conv
Out[3]: Conv2d(3, 16, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
```

**Comment by talcs** [comment]
Thank you.

If you implement the methods `get_first_conv_layer` and `set_first_conv_layer` to return and set `model.model.model[0].conv` respectively, then my sample code (from the first message) will do this conversion automatically. It will build the grayscale layer and replace the RGB one.

**Comment by Y-T-G** [comment]
And update `channels` to 1

```
In [6]: model.model.yaml["channels"] = 1
```

**Comment by Y-T-G** [comment]
And update `channels` to 1

```
In [6]: model.model.yaml["channels"] = 1
```

**Comment by Y-T-G** [comment]
@talcs You can propose a PR for the feature

**Comment by talcs** [comment]
I will try to propose a PR as soon as possible.
For now, here is my working code example:

    >>> m.model.model[0].conv
    Conv2d(3, 16, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
    >>> from convert_to_grayscale import convert_to_grayscale
    >>> layer_getter = lambda m : m.model.model[0].conv
    >>> def layer_setter(m, l):
    ...   m.model.model[0].conv = l
    ... 
    >>> convert_to_grayscale(m, layer_getter, layer_setter)
    >>> m.model.model[0].conv
    Conv2d(1, 16, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
    >>> m.model.yaml['channels'] = 1
    >>> m.export(format='onnx', imgsz=(480,640))

The function `convert_to_grayscale` is the one from my code example in the first message.

**Comment by glenn-jocher** [comment]
That approach looks fine—after swapping the first `Conv2d` and setting `m.model.yaml["channels"] = 1`, you can also sanity-check the “exact equivalence to gray duplicated into 3 channels” claim by comparing raw PyTorch forward outputs (same normalization) before exporting; and for runtime inputs, passing a 2D NumPy image (`H×W`) is supported since the loader will auto-expand grayscale to `H×W×1` (no 3-channel duplication) as documented in the [data loader reference](https://docs.ultralytics.com/reference/data/loaders/). If you open a PR, please keep it tightly scoped (e.g., an optional export-time switch) and add a small export test that asserts the exported model input shape is 1-channel while leaving default exports unchanged.
